components/Init/src/virtio_vsock.c

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright 2022, UNSW (ABN 57 195 873 179)
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+#include <stdint.h>
+#include <string.h>
+#include <autoconf.h>
+
+#include <utils/util.h>
+
+#include <camkes.h>
+#include <camkes/dataport.h>
+
+#include <sel4vm/guest_irq_controller.h>
+#include <sel4vm/boot.h>
+
+#include <sel4vmmplatsupport/drivers/virtio_vsock.h>
+
+#include <virtqueue.h>
+#include <camkes/virtqueue.h>
+
+#include <virtio/virtio_vsock.h>
+
+#include "vm.h"
+#include "virtio_vsock.h"
+#include "virtio_irq.h"
+
+static virtio_vsock_t *virtio_vsock;
+
+typedef struct virtio_vsock_cookie {
+    vm_t *vm;
+} virtio_vsock_cookie_t;
+
+/* Configured in camkes */
+#define NUM_PORTS ARRAY_SIZE(socket_layout)
+
+/*
+ * Represents the virtqueues used in a given link between
+ * two VMs.
+ */
+typedef struct conn {
+    int cid;
+    virtqueue_driver_t *send_queue;
+    virtqueue_device_t *recv_queue;
+} conn_t;
+static conn_t connections[NUM_PORTS];
+
+static void virtio_vsock_notify_free_send(virtqueue_driver_t *queue)
+{
+    void *buf = NULL;
+    uint32_t buf_size = 0, wr_len = 0;
+    vq_flags_t flag;
+    virtqueue_ring_object_t handle;
+    while (virtqueue_get_used_buf(queue, &handle, &wr_len)) {
+        while (camkes_virtqueue_driver_gather_buffer(queue, &handle, &buf, &buf_size, &flag) >= 0) {
+            /* Clean up and free the buffer we allocated */
+            camkes_virtqueue_buffer_free(queue, buf);
+        }
+    }
+}
+
+void vsock_rx_complete(virtio_emul_t *emul, char *buf, unsigned len) WEAK;
```

Comment:
Is here only one completion call per system, or can you have one per vsock?   Plus AFAIK  `WEAK` is only effective on a sumbol definition, not a declaration.